### PR TITLE
fix(socket sink): Improve DNS resolution

### LINKED ISSUE #1574: DNS resolution doesn't work on v0.7.0 with `vector` sink
I am trying to run Vector as a sidecar agent to a Rust service I built in Kubernetes, but v0.7.0 crashes on start because of a tokio spawn error related to the DNS resolution for my `vector`-type sink. v0.6.0 works without any issues, but it looks like you've overhauled your DNS resolution code in v0.7.0 with `trust_dns_resolver` and some new futures. For what it's worth, the Rust service I built is also using `trust_dns_resolver`, but with Futures v0.1.x, Tokio v0.1.x, and the default Tokio runtime, and that works perfectly.

Here's my config file:
```toml
# Set global options
data_dir = "/var/lib/vector"

# Ingest data by tailing json log files
[sources.json_logs]
type = "file"

include = ["/var/log/my-service/*.log"]
ignore_older = 86400

# Parse the data
[transforms.json_parser]
type = "json_parser"
inputs = ["json_logs"]

# Push events to Vector services
[sinks.vector]
type = "vector"
inputs = ["json_parser"]
address = "vector.vector.svc.cluster.local:9999"
```

And here's some logs I was able to capture:
```
Jan 22 18:33:57.068  TRACE vector: Parsing config. path="/etc/vector/vector.toml"
Jan 22 18:33:57.070  DEBUG vector: Validation successful. path="/etc/vector/vector.toml"
Jan 22 18:33:57.073  INFO vector: Log level "info" is enabled.
Jan 22 18:33:57.073  INFO vector: Loading config. path="/etc/vector/vector.toml"
Jan 22 18:33:57.073  TRACE vector: Parsing config. path="/etc/vector/vector.toml"
Jan 22 18:33:57.076  TRACE mio::poll: registering with poller    
Jan 22 18:33:57.076  TRACE mio::poll: registering with poller    
Jan 22 18:33:57.076  TRACE tokio_threadpool::builder: build; num-workers=2    
Jan 22 18:33:57.076  TRACE mio::poll: registering with poller    
Jan 22 18:33:57.077  INFO vector: Vector is starting. version="0.7.0" git_version="v0.7.0" released="Wed, 22 Jan 2020 14:15:59 +0000" arch="x86_64"
Jan 22 18:33:57.077  TRACE tokio_threadpool::sender: execute; count=1    
Jan 22 18:33:57.077  TRACE tokio_threadpool::pool:     -> submit external    
Jan 22 18:33:57.077  TRACE tokio_threadpool::pool: signal_work -- notify; idx=1    
Jan 22 18:33:57.077  TRACE tokio_threadpool::pool: signal_work -- spawn; idx=1    
Jan 22 18:33:57.079  TRACE tokio_threadpool::worker: Worker::sleep; worker=WorkerId(1)    
Jan 22 18:33:57.079  TRACE tokio_threadpool::worker:   sleeping -- push to stack; idx=1    
Jan 22 18:33:57.079  TRACE tokio_threadpool::worker:     -> starting to sleep; idx=1    
Jan 22 18:33:57.079  TRACE tokio_reactor: loop process - 0 events, 0.000s    
Jan 22 18:33:57.080  TRACE tokio_threadpool::pool: signal_work -- notify; idx=1    
Jan 22 18:33:57.080  TRACE tokio_threadpool::task: Task::run; state=Running    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::async_resolver::background: trust-dns resolver running    
Jan 22 18:33:57.080  TRACE tokio_threadpool::task:     -> not ready    
Jan 22 18:33:57.080  TRACE tokio_reactor: event Readable Token(4194303)    
Jan 22 18:33:57.080  TRACE tokio_reactor: loop process - 1 events, 0.000s    
Jan 22 18:33:57.080  DEBUG trust_dns_proto::xfer::dns_handle: querying: vector.vector.svc.cluster.local.krill-stage.svc.cluster.local. A    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::name_server: reconnecting: NameServerConfig { socket_addr: V4(100.64.0.10:53), protocol: Udp, tls_dns_name: None }    
Jan 22 18:33:57.080  TRACE trust_dns_resolver::name_server::connection_provider: polling response inner    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::connection_provider: connecting: Udp { socket_addr: V4(100.64.0.10:53), timeout: 5s }    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::name_server: name_server connection failure: tokio spawn error    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::name_server: reconnecting: NameServerConfig { socket_addr: V4(100.64.0.10:53), protocol: Tcp, tls_dns_name: None }    
Jan 22 18:33:57.080  TRACE trust_dns_resolver::name_server::connection_provider: polling response inner    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::connection_provider: connecting: Tcp { socket_addr: V4(100.64.0.10:53), timeout: 5s }    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::name_server: name_server connection failure: tokio spawn error    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::name_server: reconnecting: NameServerConfig { socket_addr: V4(100.64.0.10:53), protocol: Udp, tls_dns_name: None }    
Jan 22 18:33:57.080  TRACE trust_dns_resolver::name_server::connection_provider: polling response inner    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::connection_provider: connecting: Udp { socket_addr: V4(100.64.0.10:53), timeout: 5s }    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::name_server: name_server connection failure: tokio spawn error    
Jan 22 18:33:57.080  DEBUG trust_dns_resolver::name_server::name_server: reconnecting: NameServerConfig { socket_addr: V4(100.64.0.10:53), protocol: Tcp, tls_dns_name: None }    
Jan 22 18:33:57.081  TRACE trust_dns_resolver::name_server::connection_provider: polling response inner    
Jan 22 18:33:57.081  DEBUG trust_dns_resolver::name_server::connection_provider: connecting: Tcp { socket_addr: V4(100.64.0.10:53), timeout: 5s }    
Jan 22 18:33:57.081  DEBUG trust_dns_resolver::name_server::name_server: name_server connection failure: tokio spawn error    
... (continues like this for a while)
Jan 22 18:33:57.092  DEBUG trust_dns_resolver::name_server::name_server: reconnecting: NameServerConfig { socket_addr: V4(100.64.0.10:53), protocol: Udp, tls_dns_name: None }    
Jan 22 18:33:57.092  TRACE trust_dns_resolver::name_server::connection_provider: polling response inner    
Jan 22 18:33:57.092  DEBUG trust_dns_resolver::name_server::connection_provider: connecting: Udp { socket_addr: V4(100.64.0.10:53), timeout: 5s }    
Jan 22 18:33:57.092  DEBUG trust_dns_resolver::name_server::name_server: name_server connection failure: tokio spawn error    
Jan 22 18:33:57.092  DEBUG trust_dns_resolver::name_server::name_server: reconnecting: NameServerConfig { socket_addr: V4(100.64.0.10:53), protocol: Tcp, tls_dns_name: None }    
Jan 22 18:33:57.092  TRACE trust_dns_resolver::name_server::connection_provider: polling response inner    
Jan 22 18:33:57.092  DEBUG trust_dns_resolver::name_server::connection_provider: connecting: Tcp { socket_addr: V4(100.64.0.10:53), timeout: 5s }    
Jan 22 18:33:57.092  DEBUG trust_dns_resolver::name_server::name_server: name_server connection failure: tokio spawn error    
Jan 22 18:33:57.092  TRACE tokio_threadpool::notifier: Notifier::notify; id=0x7f2202a57190    
Jan 22 18:33:57.092  TRACE tokio_threadpool::pool:     -> submit external    
Jan 22 18:33:57.092  TRACE tokio_threadpool::pool: signal_work -- notify; idx=1    
Jan 22 18:33:57.092  TRACE tokio_reactor: event Readable Token(4194303)    
Jan 22 18:33:57.092  TRACE tokio_reactor: loop process - 1 events, 0.000s    
Jan 22 18:33:57.092  TRACE tokio_threadpool::worker:     -> wakeup; idx=1    
Jan 22 18:33:57.092  ERROR vector::topology: Configuration error: Sink "vector": DNS errored Unable to resolve name: proto error
```

### PR DESCRIPTION
**Pull request by LucioFranco**
This PR fixes the issues observed in #1574 and also refactors our TCP sink to handle DNS resolution internally instead of during the build phase. Since the `vector` sink also uses this implementation it has also been fixed.

### Original Issue

The issue was originally noted in #1574 and it showed that the `vector` sink would panic if set using a host name that was not an `IPv4` or `IPv6` address. The panic turned out to be produced from [`this line`] of code from `trust-dns`. This line attempts to find the current executor in thread local storage and spawn the tcp future onto that executor. Since, in our build phase we do not currently provide access to this executor in thread local storage the `spawn` here would fail and panic (which is normal behavior if you just call `tokio::spawn`). The problem here is that the build phase is never run inside the tokio runtime or with the tokio thread locals set.

Initially, this was tested to prove to not be an issue and our tests were not catching this error because we almost exclusively setup our tests with addresses that are `IPv4`s. This means this was never caught in any of our TCP nor vector tests.

[`this line`]: https://github.com/bluejekyll/trust-dns/blob/aaffb89ddaec4c4999823e3614b4bb2acba72004/crates/resolver/src/name_server/connection_provider.rs#L162

### Discovery of the Second issue

While I was thinking about this issue and how we could resolve this in the future I noticed that our TCP sink did not actually implement DNS correctly. In fact, the TCP sink would only run DNS at the beginning during build time instead of repeatedly doing it on each TCP connect. Because of this the DNS resolver never had access to the tokio executor and neither would it allow the sink to dynamically update the DNS address.

### Fixes

With both problems stated, this PR fixes both issues in one swoop by removing all custom DNS lookups during build phase and pushing the resolution to happen while the sink is running. The downsides to this is that if there is a DNS error it will take a bit longer for it to resolve. There are also no breaking changes for the user but this is mainly a bug fix and improvement to the TCP and vector sinks.

### Questions

Now this brings up a couple points related to some original issues I opened a long time ago #170 and #300. The original DNS issue would have been a non issue if we would have represented the entire application as a future. In this case since we are a future that is getting run on the tokio runtime our thread locals are always set. This is an easy footgun to hit because there are many implicit dependencies that are not described at compile time. But instead these dependencies will cause a panic at runtime and affect our users. So does this mean that we should think about possibly resolving #300? 

Fixes #1574

Signed-off-by: Lucio Franco <luciofranco14@gmail.com>

### CODE REVIEW

**Comment by binarylogic** [comment]
Thanks for the detailed write-up. We're about to prepare a 2-week tech debt project and it sounds like these issues should be seriously considered.

File: src/sinks/util/tcp.rs [hunk 1]
```diff
@@ -76,20 +76,8 @@ impl TcpSinkConfig {
...
-            .context(DNSError)?
-            .next()
-            .ok_or(Box::new(BuildError::DNSFailure {
-                address: self.address.clone(),
-            }))?;
-
-        let port = uri.port_part().ok_or(TcpBuildError::MissingPort)?.as_u16();
-        let addr = SocketAddr::new(ip_addr, port);
+        let host = uri.host().ok_or(TcpBuildError::MissingHost)?.to_string();
+        let port = uri.port_u16().ok_or(TcpBuildError::MissingPort)?;
```
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/sinks/util/tcp.rs]: done.

File: .meta/sinks/socket.toml [hunk 2]
```diff
@@ -25,4 +25,4 @@ type = "string"
 common = true
 examples = ["92.12.333.224:5000"]
 required = true
-description = "The address to bind to."
+description = "The address to bind to, this must include some <host>:<port> pair."
```
Comment (binarylogic) [inline .meta/sinks/socket.toml]: Can we not enforce this format during validation? This doesn't need to block the PR, but we should open an issue to do so if that is the correct way to handle this.

Comment (LucioFranco) [reply .meta/sinks/socket.toml]: We do this already by providing validation errors at _build_ time.

Comment (lukesteensen) [reply .meta/sinks/socket.toml]: The existing part of this is actually incorrect. We don't bind to this address, we connect to it.

Comment (LucioFranco) [reply .meta/sinks/socket.toml]: Ah yeah, we should update that.

Comment (LucioFranco) [reply .meta/sinks/socket.toml]: done in #1585

**Pull request merged**